### ISSUE TITLE
Slow resources load

### ISSUE THREAD

**Issue by talalmajali** [learntoflutter/flutter_embed_unity#32]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Unity version

- [X] I am using Unity 2022.3 LTS

### Common issues

- [X] I have checked the common issues

### Flutter version

3.22~3.24

### Description

The app freezes when loading a bulk of resources (images/prefabs) during runtime, and after running the profiler you can see it's related to File.Read when it comes to fetch the assets from the resources folder, when you export the exact same app as gradle project (without flutter) you can see all resources loading are fast, also i've tried this using Unity 2022.3.32f and 2022.3.54f

Unity 2022.3.54f embedded inside flutter
![Screenshot from 2024-12-19 19-20-47](https://github.com/user-attachments/assets/fb42d8dd-d467-44f9-84d2-ba89e5d23594)
![image](https://github.com/user-attachments/assets/a0cd3f59-87d6-4d62-abc5-165c4247ad19)

Unity 2022.3.54f just a native android export
![Screenshot from 2024-12-19 18-53-27](https://github.com/user-attachments/assets/c683df69-decf-47b6-99ef-686de0063de2)

same behavior applies to 2022.3.32f as well.

any idea how to solve this? 

### What platforms are you seeing the problem on?

Android

### Devices

Infinix HOT 10 Play (Android 10) armeabi-v7a
huawei nova 3i (Android 9) arm64

### Anything else?
Repo link: https://github.com/talalmajali/simple-saber
_No response_

**Comment by RF103T** [comment]
This problem is caused that unity's resource files are compressed twice during flutter build apk.

https://discussions.unity.com/t/extremly-slow-resources-load-file-read-on-android/890506/4
``` 
The easiest way to check this is to open apk with 7z and check the Store method, if it’s Deflate then it means it’s compressed in APK. 
```

![1](https://github.com/user-attachments/assets/eaa8211d-2e9b-4cff-95da-d8049f2dbea1)

For a quick fix, you can copy ` android.aaptOptions ` from `your_flutter_project/android/unityLibrary/build.gradle` to `your_flutter_project/android/app/build.gradle`.

When you're done, ` your_flutter_project/android/app/build.gradle ` should have a similar structure (depending on your version of unity):
```
android {

    ...

    aaptOptions {
        noCompress = ['.unity3d', '.ress', '.resource', '.obb', '.bundle', '.unityexp'] + unityStreamingAssets.tokenize(', ')
        ignoreAssetsPattern = "!.svn:!.git:!.ds_store:!*.scc:!CVS:!thumbs.db:!picasa.ini:!*~"
    }

    ...

}
```

I have not found any drawbacks and side effects yet.

**Comment by talalmajali** [comment]
@RF103T it did the trick, that you very much!! seems i have to carefully check the difference between unity normal export (app/gradle.build) and the flutter one, loading resources now is much better, i think it worth to add this to the documentation

**Comment by RF103T** [comment]
> The issue might be caused by Unity's resource files being compressed twice during flutter build apk.
> 
> https://discussions.unity.com/t/extremly-slow-resources-load-file-read-on-android/890506/4
> 
> ```
> The easiest way to check this is to open apk with 7z and check the Store method, if it’s Deflate then it means it’s compressed in APK. 
> ```
> 
> ![1](https://github.com/user-attachments/assets/eaa8211d-2e9b-4cff-95da-d8049f2dbea1)
> 
> For a quick fix, you can copy `android.aaptOptions` from `your_flutter_project/android/unityLibrary/build.gradle` to `your_flutter_project/android/app/build.gradle`.
> 
> When you're done, `your_flutter_project/android/app/build.gradle` should have a similar structure (depending on your version of unity):
> 
> ```
> android {
> 
>     ...
> 
>     aaptOptions {
>         noCompress = ['.unity3d', '.ress', '.resource', '.obb', '.bundle', '.unityexp'] + unityStreamingAssets.tokenize(', ')
>         ignoreAssetsPattern = "!.svn:!.git:!.ds_store:!*.scc:!CVS:!thumbs.db:!picasa.ini:!*~"
>     }
> 
>     ...
> 
> }
> ```
> 
> I have not found any drawbacks and side effects yet.

@jamesncl Hey, I've been testing this solution in my own project for a while, and I haven't found any obvious problems (Though it might increase the size of the installation package?). Would it be possible to consider adding it to the documentation?

**Comment by timbotimbo** [comment]
I've been using this with the other Unity plugin for a while, and haven't had any problems yet either.  
Some 3D models with textures loaded from local addressables load 1-2 seconds faster, depending on the phone used.

In my app I removed `+ unityStreamingAssets.tokenize(', ')` since I'm not using any streamingAssets.

**Comment by jamesncl** [comment]
> @jamesncl Hey, I've been testing this solution in my own project for a while, and I haven't found any obvious problems (Though it might increase the size of the installation package?). Would it be possible to consider adding it to the documentation?

@RF103T Sorry I missed this earlier, better late than never!

It's also slightly improved the load time of my own app. Nice.

I've added this to the README on the main branch, and to the examples. It won't appear in the README on pub.dev until a new version is published - I figured I'll just wait until we've tested #39 for long enough and it'll get added when I publish the production version of 1.2.6.

**Comment by RF103T** [comment]
> I noticed a changelog entry related to this in several Unity versions.
> 
> > Android: Enforcing to use Streaming Assets extensions for noCompress string in Android Gradle project to avoid problem with too many entries in the list, which may cause Gradle build fail. ([UUM-105133](https://issuetracker.unity3d.com/issues/android-build-fails-when-streamingassets-contains-998-or-more-files))
> 
> > Fixed in 2021.3.54f1, 2022.3.64f1, 6000.0.52f1, 6000.1.9f1, 6000.2.0b7, 6000.3.0a1, 7000.0.0a28
> 
> Not sure if this changed anything for us, but i'm adding it here as a hint in case any issues show up in the future.
> 
> The closed PR [#42](https://github.com/learntoflutter/flutter_embed_unity/issues/42), is also related to this topic.

This fix should be good for us as well, but only if `unityStreamingAssets` are configured correctly in `gradle.properties` according to README. It is possible to add this limitation in README and suggests that it can be optimized manually if necessary.
